### PR TITLE
[TEVA-1824] Send all create update delete events to BigQuery

### PR DESCRIPTION
**Pull request by cesidio**
## Jira ticket URL
https://dfedigital.atlassian.net/browse/TEVA-1824

## Changes in this PR:
Implementing analytics events from models, as spiked in this [PR](https://github.com/DFE-Digital/teaching-vacancies/pull/2934)

### CODE REVIEW

**Comment by martin-bangoura** [comment]
Review app deployed to <https://teaching-vacancies-review-pr-2936.london.cloudapps.digital>

File: app/models/application_record.rb [hunk 1]
```diff
@@ -1,3 +1,13 @@
...
   self.abstract_class = true
+
+  after_create { EventContext.trigger_event(:entity_created, event_data) }
+  after_update { EventContext.trigger_event(:entity_updated, event_data) unless saved_change_to_attribute?(:last_activity_at) }
+  after_destroy { EventContext.trigger_event(:entity_destroyed, event_data) }
+
+  private
+
+  def event_data
+    { entity: self.class.name }.merge(attributes.slice(*Rails.configuration.analytics[self.class.name.to_sym]))
```
Comment (cpjmcquillan) [inline app/models/application_record.rb:11]: Maybe separate 
```ruby
attributes.slice(*Rails.configuration.analytics[self.class.name.to_sym])
```
into a method with a name that's a bit more revealing?

File: app/models/application_record.rb [hunk 2]
```diff
@@ -1,3 +1,29 @@
...
+
+  def event_data
+    { entity: self.class.name }.merge(anonymised_attributes)
+  end
+
+  def anonymised_attributes
+    attributes.each_with_object({}) do |(key, value), anonymised|
+      next unless export?(key)
+
+      anonymised[key] = anonymise?(key) ? StringAnonymiser.new(value).to_s : value
```
Comment (cpjmcquillan) [inline app/models/application_record.rb:18]: I know in some places anonymised data was suffixed with `-identifier`. If @stevenleggdfe is happy we should change those to be consistent and just send the attribute key without `-identifier` everywhere

File: config/analytics.yml [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    - school_type
+    - local_authority_code
+    - group_type
+    - local_authority_within
+    - establishment_status
+  OrganisationVacancy:
+    - organisation_id
+    - vacancy_id
+  Publisher:
+    - oid
```
Comment (cpjmcquillan) [inline config/analytics.yml:72]: In other places we have anonymised things like this, and also `id` fields too

File: spec/models/application_record_spec.rb [hunk 4]
```diff
@@ -0,0 +1,23 @@
+require "rails_helper"
+
+RSpec.describe ApplicationRecord, type: :model do
+  describe "#after_create" do
+    context "when model is included in analytics.yml and has fields to anonymise" do
+      let(:model) { build(:publisher) }
+
+      let(:anonymised_attributes) do
+        model.attributes.merge(entity: "Publisher",
+                               "email" => StringAnonymiser.new(model.email).to_s,
```
Comment (cpjmcquillan) [inline spec/models/application_record_spec.rb:10]: You can use `anonymised_form_of` in specs

**Review by cpjmcquillan — APPROVED** [review]
Couple of small comments (mainly questions/decisions around consistency with the rest of the application that can be addressed in a separate PR) - but I love it!

**Review by cpjmcquillan — APPROVED** [review]

**Pull request merged**